README.md
# Flutter Romantic Expandable Card with folding effect

A Flutter package to create a Romantic Expandable Card with folding effect. It will be useful for your awesome app.
This widget is created to be standalone with any widgets, helps you able to add to any complex design.

This is a present from Romantic Project. More from [Romantic Developer](https://pub.dev/publishers/romanticdeveloper.com/packages)

![Demo](./demo.gif)

### Example

```dart
import 'package:flutter/material.dart';
import 'package:flutter_folding_card/flutter_folding_card.dart';

void main() => runApp(MyApp());

const _kImageUrls = [
  "https://scontent-hkg4-1.xx.fbcdn.net/v/t1.6435-9/240669493_3093165274342302_7696944317595605164_n.jpg?_nc_cat=107&ccb=1-5&_nc_sid=730e14&_nc_ohc=QRmRTo3bJOQAX_u8ELe&_nc_ht=scontent-hkg4-1.xx&oh=374ca65aaff7a46dfd4d2f14aff72b11&oe=614F64BC",
  "https://scontent-hkg4-2.xx.fbcdn.net/v/t1.6435-9/240758821_3092745557717607_1758983175902930666_n.jpg?_nc_cat=109&ccb=1-5&_nc_sid=730e14&_nc_ohc=z0srgae32A4AX_oURvc&_nc_ht=scontent-hkg4-2.xx&oh=e981297264aca64a25aaf285259f41d7&oe=614CDEA1",
  "https://scontent-hkg4-1.xx.fbcdn.net/v/t1.6435-9/240624734_3092487924410037_2111143168440995076_n.jpg?_nc_cat=107&ccb=1-5&_nc_sid=730e14&_nc_ohc=TZukbOFjxowAX8IYthH&_nc_ht=scontent-hkg4-1.xx&oh=59dd511589b6dbdea97a3b478b5d629f&oe=614D9448",
  "https://scontent-hkg4-1.xx.fbcdn.net/v/t1.6435-9/s640x640/237446804_3091585131166983_1116147550483070313_n.jpg?_nc_cat=105&ccb=1-5&_nc_sid=8bfeb9&_nc_ohc=5QCW7JFNQoAAX-Suuyv&_nc_ht=scontent-hkg4-1.xx&oh=ca4d078c57e9b95815d6facb38bef7db&oe=614C87E3",
];

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final itemCount = 3;
  final foldOutList = <bool>[false, false, false];

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: Text('Example'),
          actions: [
            IconButton(
              onPressed: () {
                setState(() {
                  for (var i = 0; i < foldOutList.length; ++i) {
                    foldOutList[i] = false;
                  }
                });
              },
              icon: Icon(Icons.cleaning_services_sharp),
            ),
          ],
        ),
        backgroundColor: Colors.white,
        body: ListView.builder(
          itemBuilder: (context, index) {
            return Padding(
              padding: const EdgeInsets.only(left: 22.0, right: 22),
              child: FoldingCard(
                foldOut: foldOutList[index],
                curve: foldOutList[index] == true
                    ? Curves.easeInCubic
                    : Curves.easeOutCubic,
                duration: Duration(milliseconds: 1400),
                coverBackground: ElevatedButton(
                  onPressed: () {
                    setState(() {
                      foldOutList[index] = true;
                    });
                  },
                  child: Text(
                    'This is a sample coverBackground, click on it to fold in.',
                    textAlign: TextAlign.center,
                  ),
                ),
                expandedCard: index == 1
                    ? Stack(
                        children: [
                          Image.network(
                            _kImageUrls[3],
                            fit: BoxFit.fitWidth,
                            width: MediaQuery.of(context).size.width,
                            alignment: Alignment.topCenter,
                          ),
                          Center(
                            child: ElevatedButton(
                              onPressed: () {},
                              child: Text(
                                'This is a other sample for expandedCard.',
                              ),
                            ),
                          )
                        ],
                      )
                    : Image.network(
                        _kImageUrls[1],
                        fit: BoxFit.cover,
                        width: MediaQuery.of(context).size.width,
                        alignment: Alignment.topCenter,
                      ),
                cover: ElevatedButton(
                  style: ButtonStyle(
                    padding: MaterialStateProperty.all(EdgeInsets.zero),
                  ),
                  onPressed: () {
                    setState(() {
                      foldOutList[index] = false;
                    });
                  },
                  child: Image.network(
                    _kImageUrls[2],
                    fit: BoxFit.fitWidth,
                    width: MediaQuery.of(context).size.width,
                    alignment: Alignment.topCenter,
                  ),
                ),
                foldingHeight: 100,
                expandedHeight: 300,
              ),
            );
          },
          itemCount: itemCount,
        ),
      ),
    );
  }
}
```

### Development environment

```
[✓] Flutter (Channel stable, 3.0.1, on macOS 12.3.1 21E258 darwin-x64, locale en-VN)
    • Flutter version 3.0.1 at ~/fvm/versions/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision fb57da5f94 (5 days ago), 2022-05-19 15:50:29 -0700
    • Engine revision caaafc5604
    • Dart version 2.17.1
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at ~/Library/Android/sdk
    • Platform android-31, build-tools 30.0.3
    • ANDROID_HOME = ~/Library/Android/sdk
    • ANDROID_SDK_ROOT = ~/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.40.0 
```

CHANGELOG.md
## [1.1.0] - 25/05/2022

* Migrate to flutter 3.0.1.

## [1.0.0+1] - 03/09/2021

* Edit description.

## [1.0.0] - 30/08/2021

* First release version.


lib/src/folding_card.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';

/// It helps you in creating a Romantic Folding Card.
/// This widget is created to be able to standalone with any views, helps you able to add it into any complex design.
class FoldingCard extends StatefulWidget {
  final Widget expandedCard;
  final double expandedHeight;
  final double foldingHeight;
  final Widget cover;
  final Widget? coverBackground;
  final bool foldOut;
  final Function(double value, AnimationStatus status)? listener;
  final Curve curve;
  final Duration duration;
  final BoxDecoration? foldingShadow;
  final bool foldingShadowVisible;

  /// Create a Romantic Folding Card.
  ///
  /// This widget will base on [foldOut] to decide to run the folding animation. The final widget after folded out is a [Column] includes [expandedCard] and [cover].
  ///
  /// The [expandedCard] is fixed by [expandedHeight], and the [cover] & [coverBackground] is fixed by [foldingHeight].
  ///
  /// [expandedHeight] & [foldingHeight] it's used to calculate the [_FoldingCardState._foldingCount].
  ///
  /// [foldingHeight] is height of each folding parts.
  ///
  /// During animating, when the [cover] is folding at 50% each of cycles, it will be turned into the [coverBackground].
  /// And it has a [foldingShadow] that placed backward of [coverBackground] and [cover].
  /// There is a default [foldingShadow], you can custom it also if you don't want to show it, set [foldingShadowVisible] to false, default is true.
  ///
  /// You can also custom the overall animation by using [curve] & [duration].
  const FoldingCard({
    Key? key,
    this.foldOut: false,
    this.listener,
    this.curve: Curves.linear,
    this.duration: const Duration(milliseconds: 1200),
    required this.cover,
    this.coverBackground,
    required this.expandedCard,
    required this.expandedHeight,
    required this.foldingHeight,
    this.foldingShadow,
    this.foldingShadowVisible: true,
  })  : assert(expandedHeight ~/ foldingHeight >= 1),
        super(key: key);

  @override
  _FoldingCardState createState() => _FoldingCardState();
}

class _FoldingCardState extends State<FoldingCard>
    with TickerProviderStateMixin {
  late final AnimationController _controller;
  late final AnimationController _controllerShadow;

  late TweenSequence<Alignment> _coverAlignmentTween;
  late TweenSequence<Offset> _coverTranslateTween;
  late int _foldingCount;
  late TweenSequence<double> _shadowOpacityTween;
  late TweenSequence<double> _foldingOutBottomMarginTween;
  late TweenSequence<double> _foldingInTopMarginTween;

  @override
  void initState() {
    _controller =
        AnimationController(vsync: this, duration: widget.duration, value: 1.0);
    _controller.addListener(() {
      if (widget.listener != null) {
        widget.listener!(_controller.value, _controller.status);
      }
      setState(() {});
    });
    _controllerShadow =
        AnimationController(vsync: this, duration: widget.duration);
    _controllerShadow.addListener(() {
      setState(() {});
    });
    _initTweens();
    super.initState();
  }

  void _initTweens() {
    _foldingCount = (widget.expandedHeight / widget.foldingHeight).round();
    if (_foldingCount * widget.foldingHeight > widget.expandedHeight) {
      _foldingCount--;
    } else if (_foldingCount * widget.foldingHeight < widget.expandedHeight) {
      _foldingCount++;
    }
    _coverAlignmentTween = TweenSequence(List.generate(_foldingCount, (index) {
      var a = index % 2 == 0 ? Alignment.bottomCenter : Alignment.topCenter;
      return TweenSequenceItem(
        tween: Tween(begin: a, end: a),
        weight: _weightPerPage,
      );
    }));
    _coverTranslateTween = TweenSequence(List.generate(
      _foldingCount,
      (index) {
        var y = index ~/ 2 * (-widget.foldingHeight * 2);
        return TweenSequenceItem(
          tween: Tween(begin: Offset(0, y), end: Offset(0, y)),
          weight: _weightPerPage,
        );
      },
    ));
    _shadowOpacityTween = TweenSequence([
      TweenSequenceItem(
        tween: Tween(begin: 0.0, end: 1.0),
        weight: 0.1,
      ),
      TweenSequenceItem(
        tween: Tween(begin: 1.0, end: 1.0),
        weight: 0.8,
      ),
      TweenSequenceItem(
        tween: Tween(begin: 1.0, end: 0.0),
        weight: 0.1,
      ),
    ]);
    _foldingOutBottomMarginTween = TweenSequence([
      TweenSequenceItem(
        tween: Tween(begin: 0.0, end: widget.foldingHeight),
        weight: 0.2,
      ),
      TweenSequenceItem(
        tween: Tween(begin: widget.foldingHeight, end: widget.foldingHeight),
        weight: 0.7,
      ),
      TweenSequenceItem(
        tween: Tween(begin: widget.foldingHeight, end: 0.0),
        weight: 0.1,
      ),
    ]);
    _foldingInTopMarginTween = TweenSequence([
      TweenSequenceItem(
        tween: Tween(begin: widget.foldingHeight, end: 0.0),
        weight: _weightPerPage / 2,
      ),
      TweenSequenceItem(
        tween: Tween(begin: 0.0, end: widget.foldingHeight),
        weight: 1 - _weightPerPage,
      ),
      TweenSequenceItem(
        tween: Tween(begin: widget.foldingHeight, end: 0.0),
        weight: _weightPerPage / 2,
      ),
    ]);
  }

  double get _weightPerPage => 1.0 / _foldingCount;

  @override
  void dispose() {
    _controller.dispose();
    _controllerShadow.dispose();
    super.dispose();
  }

  /// This logic is used to update exactly the current animation state. If not, there will be caused a bug if this [FoldingCard] is placed in a [ListView].
  @override
  void didChangeDependencies() {
    _controllerShadow.value = 0;
    if (widget.foldOut) {
      _controller.value = 0;
    } else {
      _controller.value = 1;
    }
    super.didChangeDependencies();
  }

  @override
  void didUpdateWidget(covariant FoldingCard oldWidget) {
    _initTweens();

    if (oldWidget.foldOut != widget.foldOut) {
      if (widget.foldOut) {
        _controllerShadow.forward(from: 0);
        if (_controller.value > 0) _controller.reverse(from: 1);
      } else {
        _controllerShadow.reverse(from: 1.0);
        if (_controller.value < 1) _controller.forward(from: 0);
      }
    }
    super.didUpdateWidget(oldWidget);
  }

  @override
  Widget build(BuildContext context) {
    var foldingInValue =
        widget.curve.transform(_controller.value).clamp(0.0, 1.0);
    var needBackground = foldingInValue >= _weightPerPage / 2;
    var deltaYWithTrueExpandedHeight = widget.expandedHeight +
        widget.foldingHeight -
        (_foldingCount + 1) * widget.foldingHeight;
    var expandedHeightFactor = math.max(
        widget.foldingHeight / widget.expandedHeight,
        ((1.0 - foldingInValue) * 1.22).clamp(0.0, 1.0));
    var foldingOutValue = 1 - foldingInValue;
    var coverTranslateY = widget.foldingHeight * (_foldingCount - 1) +
        deltaYWithTrueExpandedHeight;
    var foldingRotateRad = foldingInValue * -math.pi * _foldingCount;
    var coverFoldingRotateRad = _foldingCount % 2 != 0 ? math.pi : 0.0;

    return Padding(
      padding: EdgeInsets.only(
        bottom: _foldingOutBottomMarginTween.transform(foldingOutValue),
      ),
      child: Stack(
        fit: StackFit.passthrough,
        children: [
          Column(
            children: [
              ClipRect(
                child: Align(
                  alignment: Alignment.topCenter,
                  heightFactor: expandedHeightFactor,
                  child: SizedBox(
                    height: widget.expandedHeight,
                    child: widget.expandedCard,
                  ),
                ),
              ),
              Stack(
                children: [
                  SizedBox(
                    height: _foldingInTopMarginTween.transform(foldingInValue),
                  ),
                  if (widget.foldingShadowVisible)
                    Opacity(
                      opacity: _shadowOpacityTween.transform((foldingOutValue)),
                      child: Container(
                        foregroundDecoration: widget.foldingShadow ??
                            BoxDecoration(
                              boxShadow: [
                                BoxShadow(
                                  color: Colors.black54,
                                  blurRadius: widget.foldingHeight,
                                  spreadRadius: widget.foldingHeight * 0.5,
                                ),
                              ],
                            ),
                      ),
                    ),
                ],
              ),
            ],
          ),
          Transform(
            transform: Matrix4.identity()..translate(0.0, coverTranslateY),
            child: Transform(
              transform: Matrix4.identity()
                ..translate(
                    0.0,
                    _coverTranslateTween.transform(foldingInValue).dy -
                        (needBackground ? deltaYWithTrueExpandedHeight : 0)),
              child: Transform(
                alignment: _coverAlignmentTween.transform(foldingInValue),
                transform: Matrix4.identity() //
                  ..setEntry(3, 2, 0.001)
                  ..rotateX(foldingRotateRad),
                child: Transform(
                  alignment: _coverAlignmentTween.transform(foldingInValue),
                  transform: Matrix4.identity() //
                    ..rotateX(math.pi),
                  child: Transform(
                    alignment: Alignment.center,
                    transform: Matrix4.identity() //
                      ..rotateX(math.pi),
                    child: Stack(
                      children: [
                        SizedBox(
                          child: widget.cover,
                          width: double.infinity,
                          height: widget.foldingHeight,
                        ),
                        if (needBackground && widget.coverBackground != null)
                          Transform(
                            alignment: Alignment.center,
                            transform: Matrix4.identity() //
                              ..rotateX(coverFoldingRotateRad),
                            child: SizedBox(
                              child: widget.coverBackground!,
                              width: double.infinity,
                              height: widget.foldingHeight,
                            ),
                          ),
                      ],
                    ),
                  ),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_folding_card/flutter_folding_card.dart';

void main() => runApp(MyApp());

const _kImageUrls = [
  "assets/sample_1.jpg",
  "assets/sample_2.jpg",
  "assets/sample_3.jpg",
];

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final itemCount = 3;
  final foldOutList = <bool>[false, false, false];

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: Text('Example'),
          actions: [
            IconButton(
              onPressed: () {
                setState(() {
                  for (var i = 0; i < foldOutList.length; ++i) {
                    foldOutList[i] = false;
                  }
                });
              },
              icon: Icon(Icons.cleaning_services_sharp),
            ),
          ],
        ),
        backgroundColor: Colors.white,
        body: ListView.builder(
          itemBuilder: (context, index) {
            return Padding(
              padding: const EdgeInsets.only(left: 22.0, right: 22),
              child: FoldingCard(
                foldOut: foldOutList[index],
                curve: foldOutList[index] == true
                    ? Curves.easeInCubic
                    : Curves.easeOutCubic,
                duration: Duration(milliseconds: 1400),
                coverBackground: ElevatedButton(
                  onPressed: () {
                    setState(() {
                      foldOutList[index] = true;
                    });
                  },
                  child: Text(
                    'This is a sample coverBackground, click on it to fold in.',
                    textAlign: TextAlign.center,
                  ),
                ),
                expandedCard: index == 1
                    ? Stack(
                        children: [
                          Image.asset(
                            _kImageUrls[0],
                            fit: BoxFit.fitWidth,
                            width: MediaQuery.of(context).size.width,
                            alignment: Alignment.topCenter,
                          ),
                          Center(
                            child: ElevatedButton(
                              onPressed: () {},
                              child: Text(
                                'This is a other sample for expandedCard.',
                              ),
                            ),
                          )
                        ],
                      )
                    : Image.asset(
                        _kImageUrls[1],
                        fit: BoxFit.cover,
                        width: MediaQuery.of(context).size.width,
                        alignment: Alignment.topCenter,
                      ),
                cover: ElevatedButton(
                  style: ButtonStyle(
                    padding: MaterialStateProperty.all(EdgeInsets.zero),
                  ),
                  onPressed: () {
                    setState(() {
                      foldOutList[index] = false;
                    });
                  },
                  child: Image.asset(
                    _kImageUrls[2],
                    fit: BoxFit.fitWidth,
                    width: MediaQuery.of(context).size.width,
                    alignment: Alignment.topCenter,
                  ),
                ),
                foldingHeight: 100,
                expandedHeight: 300,
              ),
            );
          },
          itemCount: itemCount,
        ),
      ),
    );
  }
}
